File: raw_data_processing/EV_File_Using_Template.py
###############################################################################

# EV script 010 - New EV file using template, add data, create lines

# Example Echoview COM script originally submitted to the Echoview forum by Jeremy Holden

# Downloaded from www.echoview.com

# For assistance, contact Echoview support <support@echoview.com>

###############################################################################

# Adapted January 27th 2023 by Delphine Mossman

# The goal of this script is to automate the creation and import of new Echoview files for each day of AZFP data
# %%

# Required libraries
import win32com.client
import os
import re

# %%
# Define some variables

homedir = "H:/dm1679/Data/Glider Data/"
dep_name = input('Enter the full name of the deployment: ')

template = os.path.join(homedir,'AZFP_Fish_Template_2023.EV')
# template = os.path.join(homedir,'AZFP_Zooplankton_Template_2023.EV')

workdir = os.path.join(homedir,dep_name,'')

# workdir will have to be changed for each new glider deployment

# Open EV connection

EvApp = win32com.client.Dispatch("EchoviewCom.EvApplication")



# Define the year and month to work with, and where the day files are found
AZFPyear = dep_name[5:9]

basedir = os.path.join(workdir,'AZFP Data','')

for folder in next(os.walk(basedir))[1]:
# for folder in ['202410']:
    AZFPmonth = folder[-2:]
    daydir = os.path.join(workdir, 'AZFP Data',str(AZFPyear) + str(AZFPmonth),'')

    for day in os.listdir(daydir):
    # for day in ['21']:
        # Define where the AZFP and GPS/roll/pitch files are
        AZFPdir = os.path.join(daydir,day,'')

        CSVdir = os.path.join(workdir,'Echoview CSV Import Files','')
        # Open a new EV file using the template
        EvFile = EvApp.NewFile(template)

        # Add all AZFP files from within AZFPdir

        ## Specify file extension for AZFP data

        extension = '[.]01.$'

        ## Get list of files to add

        myfiles = [file for file in os.listdir(AZFPdir) if re.search(extension, file)]

        ## Create loop to add all AZFP data to EV file

        for file in myfiles:

            addfile ="".join([AZFPdir,file])

            EvFile.Filesets.Item(0).Datafiles.Add(addfile)

        # Import the calibration file

        EvFile.Filesets(0).SetCalibrationFile(os.path.join(workdir, 'AZFP Data',dep_name+'.ecs'))

        # Add the CSV files into their proper tabs

        ## Specify file extension for CSV data

        extension = 'csv'

        ## Get list of files to add

        myfiles = [file for file in os.listdir(CSVdir) if file.lower().endswith(extension)]

        ## Add into proper tabs; order is always GPS, then pitch, then roll

        ### GPS
        addfile = "".join([CSVdir,myfiles[0]])
        EvFile.Filesets.Item(1).Datafiles.Add(addfile)

        ### Pitch
        addfile = "".join([CSVdir,myfiles[1]])
        EvFile.Filesets.Item(2).Datafiles.Add(addfile)

        ### Roll
        addfile = "".join([CSVdir,myfiles[2]])
        EvFile.Filesets.Item(3).Datafiles.Add(addfile)

        ### Depth (replace M69_DEPTH line)

        # input('Go open and close the 38 kHz echogram, then hit Enter to continue.')

        #### Get the M69_DEPTH line
        EvLine = EvFile.Lines.FindByName("M69_DEPTH")

        #### Import the data.depth.evl file; default name for a line is 'data'
        addfile = "".join([CSVdir,'data.depth.evl'])
        EvFile.Import(addfile)

        #### Get the new depth line
        depth = EvFile.Lines.FindByName("data")

        #### Replace the M69_DEPTH line with the depth line
        EvLine.OverwriteWith(depth)

        #### Delete the orphan depth line
        EvFile.Lines.Delete(depth)

        ### Find bottom line
        # EvFile.Properties.LinePick.Algorithm = 2
        # EvFile.Properties.LinePick.StartDepth = 15
        # EvFile.Properties.LinePick.StopDepth = 45
        # EvFile.Properties.LinePick.MinSv = -70
        # EvFile.Properties.LinePick.UseBackstep = True
        # EvFile.Properties.LinePick.DiscriminationLevel = -50.0
        # EvFile.Properties.LinePick.BackstepRange = 0

        # myvar = EvFile.Variables.FindByName("AZFP: Sv pings (averaged) T1")
        # oldline = EvFile.Lines.FindByName("Editable bottom")

        # newline = EvFile.Lines.CreateLinePick(myvar,True)
        # newline = EvFile.Lines.FindByName("Line 1")

        # oldline.OverwriteWith(newline)
        # EvFile.Lines.Delete(newline)

        # Create a folder for the month (if it does not already exist) and define the file name
        folder_path = os.path.join(workdir,'Echoview Files',str(AZFPyear) + str(AZFPmonth.zfill(2)),'')
        if not os.path.exists(folder_path):
            os.makedirs(folder_path)

        myname = os.path.join(folder_path,'',dep_name+'_Echogram_' + day + '.EV')

        # Save the file
        EvFile.SaveAs(myname)

        # Close the file
        EvApp.CloseFile(EvFile)
    


# for day in os.listdir(daydir):
# # for day in ['20']:    
#     myname = workdir + \
#     'Echoview Files/' + \
#     str(AZFPmonth.zfill(2)) + '/' + \
#     day + '/' + \
#     'ru28-20220520T1425_Echogram_' + \
#     day + '.ev'
    
#     EvFile = EvApp.OpenFile(myname)
    
#     ### Create bottom line

#     #### Set various properties of the line

    
#     # Save the file
#     EvFile.SaveAs(myname)

#     # Close the file
#     EvApp.CloseFile(EvFile)
    

# Once you are all done, quit out of Echoview entirely
EvApp.Quit()

# %%

File: raw_data_processing/EV_File_Using_Template.R
rm(list = ls())

library(tidyverse)
library(RDCOMClient)
library(R.utils)
library(tcltk)

sourceDirectory(
  "H:/dm1679/Code/R_Functions",
  modifiedOnly = F
)

## Setup
glider_dep = choose_directory(caption = "Select the glider deployment directory folder")

# Template needs to be set to match the AZFP (3 frequency fish or 4 frequency zoop)
template = choose.files(default = "C:/Users/Delphine/Box/Glider Data/AZFP*Template*.EV",
                        caption = "Select the EV template file to use") %>%
  gsub("\\\\","/",.) # Windows file selection weirdness

EVAppObj = COMCreate('EchoviewCom.EvApplication')

azfp_dep = paste0(glider_dep, "/AZFP Data/") # Raw AZFP data directory, in folders by month and day
CSV_dir = paste0(glider_dep, "/Echoview CSV Import Files/") # Raw AZFP data directory, in folders by month

## File creation
# Create an Echoview file from template for each day of the deployment, import
# the AZFP and glider data, and save

for(folder in list.dirs(azfp_dep, recursive = F)) {
  
  for(daily in list.dirs(folder, recursive = F)) {
    
    day = substr(daily, nchar(daily)-1, nchar(daily))
    
    # Create a new file from the template
    EVFile = EVAppObj$NewFile(template)
    
    # Import the calibration file
    EVFile[["Filesets"]][[0]]$SetCalibrationFile(paste0(azfp_dep, substring(glider_dep, regexpr("ru[0-9]{2}-*", glider_dep)), ".ecs"))
    
    # Get a list of all the AZFP files in the folder
    AZFP_files = list.files(daily, pattern = "*[.]01[A-Z]$", full.names = T)
    
    AZFP_fileset = EVFile[["Filesets"]]$FindByName('AZFP')
    
    # Then import the files one by one
    for(i in AZFP_files) {
      AZFP_fileset[["DataFiles"]]$Add(i)
    }
    
    # Get a list of the glider CSV files for GPS/pitch/roll
    CSV_files = list.files(CSV_dir, pattern = "*.csv$", full.names = T)
    
    # Then import the files in order: GPS, then pitch, then roll
    ### GPS
    GPS_fileset = EVFile[["Filesets"]]$FindByName('GPS')
    GPS_fileset[["DataFiles"]]$Add(CSV_files[1])
    
    ### Pitch
    Pitch_fileset = EVFile[["Filesets"]]$FindByName('Pitch')
    Pitch_fileset[["DataFiles"]]$Add(CSV_files[2])
    
    ### Roll
    Roll_fileset = EVFile[["Filesets"]]$FindByName('Roll')
    Roll_fileset[["DataFiles"]]$Add(CSV_files[3])
    
    ### Depth (replace M69_DEPTH line)
    
    # Get the M69_DEPTH line
    EVLine = EVFile[["Lines"]]$FindByName("M69_DEPTH")
    
    # Import the data.depth.evl file; default name for a line is 'data'
    EVFile$Import(paste0(CSV_dir, "data.depth.evl"))
    
    # Get the new depth line
    depth = EVFile[["Lines"]]$FindByName("data")
    
    # Replace the M69_DEPTH line with the depth line
    EVLine$OverwriteWith(depth)
    
    # Delete the orphan depth line
    EVFile[["Lines"]]$Delete(depth)
    
    # Create a folder for the month (if it does not already exist) and define the file name
    create_dir(gsub("AZFP Data", "Echoview Files", folder))
    
    EV_file_name = paste0(gsub("AZFP Data", "Echoview Files", folder),"/",
                          substring(glider_dep, regexpr("ru[0-9]{2}-*", glider_dep)),
                          "_Echogram_",day,".EV")
    
    # Save the file
    EVFile$SaveAs(EV_file_name)
    
    
    # Close the file
    EVFile$Close()
    
  }
}

EVAppObj$Quit()
